src/ir_Argo.cpp

```diff
@@ -505,4 +1748,98 @@ bool IRrecv::decodeArgo(decode_results *results, uint16_t offset,
+/// @note This decoder is separate from @c decodeArgo to maintain backwards
+///       compatibility. Contrary to WREM2, this expects a footer and gap!
+bool IRrecv::decodeArgoWREM3(decode_results *results, uint16_t offset,
+                             const uint16_t nbits,
+                             const bool strict) {
+  if (strict
+      && nbits != kArgo3AcControlStateLength * 8
+      && nbits != kArgo3ConfigStateLength * 8
+      && nbits != kArgo3iFeelReportStateLength * 8
+      && nbits != kArgo3TimerStateLength * 8) {
+    return false;
+  }
+
+  const uint16_t kArgoOverhead = 3;
+  uint16_t bytesRead = matchGeneric(results->rawbuf + offset, results->state,
+                  results->rawlen - offset, nbits,
+                  kArgoHdrMark, kArgoHdrSpace,
+                  kArgoBitMark, kArgoOneSpace,
+                  kArgoBitMark, kArgoZeroSpace,
+                  kArgoBitMark, kArgoGap,  // difference vs decodeArgo
+                  true, _tolerance, kArgoOverhead,
+                  false);
+  if (!bytesRead ||
+      !IRArgoAC_WREM3::isValidWrem3Message(results->state, nbits, strict)) {
+    return false;
+  }
+
+  // Success: Matched ARGO protocol and WREM3-model
+  // Note that unfortunately decode_type does not allow to persist model...
+  // so we will be re-detecting it later :)
+  results->decode_type = decode_type_t::ARGO;
+  results->bits = nbits;
+  // No need to record the state as we stored it as we decoded it.
+  // As we use result->state, we don't record value, address, or command as it
+  // is a union data type.
+  return true;
+}
+
+/// @brief Detects if an ARGO protocol message is a WREM-3 sub-type (model)
+/// @param state The raw IR decore state
+/// @param nbits The length of @c state **IN BITS**
+/// @param strict Whether to perform strict matching (incl. checksum)
+/// @return True if the message is a WREM-3 one
+bool IRArgoAC_WREM3::isValidWrem3Message(const uint8_t state[],
+                                         const uint16_t nbits, bool strict) {
+  auto stateLength = std::min(static_cast<uint16_t>(std::ceil(nbits / 8.0)),
+                              kStateSizeMax);
+
+  if (strict && !IRArgoAC_WREM3::hasValidPreamble(state, stateLength)) {
+    return false;
+  }
+
+  auto messageType = IRArgoACBase<ArgoProtocolWREM3>::getMessageType(state,
+    stateLength);
+
+  switch (messageType) {
+    case argoIrMessageType_t::AC_CONTROL :
+      if (nbits != kArgo3AcControlStateLength * 8) { return false; }
+      break;
+  case argoIrMessageType_t::CONFIG_PARAM_SET:
+      if (nbits != kArgo3ConfigStateLength * 8) { return false; }
+      break;
+  case argoIrMessageType_t::TIMER_COMMAND:
+      if (nbits != kArgo3TimerStateLength * 8) { return false; }
+      break;
+    case argoIrMessageType_t::IFEEL_TEMP_REPORT:
+      if (nbits != kArgo3iFeelReportStateLength * 8) { return false; }
+      break;
+    default:
+      return false;
+  }
```

Comment:
> That will reduce the resultant code size, and improve readablity. e.g. `statelength`.

Good point - will simplify (agree with the readability part of the comment).
Must add though, that I'd be very surprised it does anything positive to the resulting code size. 
Compiler should be perfectly capable of optimizing it on its own.

----

> Also, this really is a "strict" condition IMHO.

Well, yes and no.
That is, it is about the only thing separating `WREM2` from `WREM3` (and WREM3 detection must run before WREM2 non-strict, as it is a swallow-all).
While there's indeed a `getMessageType()` doing the detect, it only takes 2 **bits** of input, so a chance of a false-positive is way high. Aka. I consider "checksum" strict check, but anything else... I deliberately wanted to fall through to WREM2 and not get handled here.

That said, I only use the detect in strict mode anyway, so the point I'm making above is a bit moot and I could equally well remove 'strict' param entirely...
Leaving as-is for now, let me know where you'd like to take it.

Reply:
My unwritten policy for `strict` is, if it is in some way protocol-format related and can cause it to abort decoding, encapsulate it in `strict`. The purpose is to allow more loose matching with `strict` set to `false`, and/or code re-use where the protocol timings etc are compatible, but the length or structure of the data differs in a different protocol.
i.e. Maximum potential for code-reuse to reduce library binary size etc.

Reply:
Please note though that `WREM2` and `WREM3` are variants (models) of the same `ARGO` IR protocol. 
The difference is in message decoding and parsing, so in the case you describe:
>... but the length or structure of the data differs in a different protocol....

...there's already a `decodeArgo(..., strict=false)` which would accomplish that result just fine.
The `WREM3` flavor was meant signify this is semantically a different sub-proto.

If I make the length-check non-strict, then, arguably, `decodeArgoWREM3(..., strict=false)` would be almost no different from `decodeArgo(..., strict=false)`... which makes it a bit redundant IMHO (but arguably consistent)

My preferences (from most-preferred):
1. Leave as is (possibly less semantically consistent, but the most versatile impl.)
2. Remove `strict` param from `decodeArgoWREM3` altogether
3. Make the length non-strict (it has a potential to clash with wrem2, but since I'm not calling it this way anyway... I guess the user would know what they're after). 
I.e. when you call it with `strict=false` a `true` response no longer means there's any confidence in it really being a WREM3.

Your pick? :)

Reply:
Users typically can't call the `decodeProtocol()` routines directly. So, it's just us "internal" users.
Practically every `decodeProtocol()` has a `strict`, so I'd like to keep that calling convention.
Of the options, **3.** sounds best to me. It is almost always a given if a `decodeProtocol()` is called with `strict=false` that it's not a confident match. It assumes the caller knows what they are doing.